python/ml4ir/base/model/scoring/monte_carlo_scorer.py

```diff
@@ -99,12 +91,171 @@ def __init__(
+            value_shape = tf.shape(value)
+            flattened_shape = tf.concat([[value_shape[0] * value_shape[1]], value_shape[2:]], axis=0)
+            flattened_inputs[key] = tf.reshape(value, flattened_shape)
+
+        # Compute scores for all masked inputs in one go
+        all_scores = super().call(flattened_inputs, training=False)[self.output_name]
+        return all_scores
+
+    def reshape_and_normalize(self, all_scores, mask_count, batch_size):
+        """
+        Reshape the scores tensor to separate mask and batch dimensions, then normalize the scores.
+
+        Parameters
+        ----------
+        all_scores : tf.Tensor
+            Tensor containing scores computed for all masked inputs. The shape is expected to be [mask_count * batch_size, ...].
+        mask_count : int
+            The number of different masks applied. This corresponds to the number of Monte Carlo trials.
+        batch_size : int
+            The size of the batch of inputs.
+
+        Returns
+        -------
+        tf.Tensor
+            The reshaped and normalized scores tensor. The shape of the tensor will be [batch_size, ...].
+        """
+        all_scores_shape = tf.shape(all_scores)
+        all_scores = tf.reshape(all_scores, tf.concat([[mask_count, batch_size], all_scores_shape[1:]], axis=0))
+
+        # Aggregate the scores across the mask dimension
+        all_scores = tf.reduce_sum(all_scores, axis=0)
+        all_scores = tf.divide(all_scores, self.monte_carlo_fixed_trials_count)
+        return all_scores
+
+    def deterministic_call(self, inputs: Dict[str, tf.Tensor], training=None):
+        """
+        Compute scores for input features using a deterministic approach with fixed masks.
+
+        Parameters
+        ----------
+        inputs : dict of tf.Tensor
+            Dictionary containing input feature tensors. Each key represents the feature name, and each value is a tensor of shape [batch_size, feature_length].
+        training : bool, optional
+            Indicator of whether the model is in training mode. This parameter is passed to the scoring method to control behavior specific to training or inference. Default is None.
+
+        Returns
+        -------
+        dict of tf.Tensor
+            A dictionary containing the computed scores. The key is `self.output_name`, and the value is a tensor with the shape [batch_size, ...].
+        """
+        batch_size = tf.shape(list(inputs.values())[0])[0]
+        mask_count = len(self.fixed_mask)
+
+        # Apply fixed masks
+        masked_inputs_list = self.mask_inputs(inputs, mask_count)
+        # score with the fixed mask
+        all_scores = self.score_with_fixed_mask(inputs, masked_inputs_list, training)
+        # Reshape all_scores to separate the mask dimension and the original batch dimension
+        all_scores = self.reshape_and_normalize(all_scores, mask_count, batch_size)
+        return {self.output_name: all_scores}
```

Comment:
@mohazahran The deterministic call is breaking the definition of what MCScorer does and is leaking code from the QueryFeatureMask. Please refactor.